### PR TITLE
eframe: Don't show window until after initialization

### PR DESCRIPTION
**Pull request by LoganDark**
Shortens #1802, but does not completely solve it. This looks to be a step in the right direction, though, and eframe should have been doing this anyway.

Marked as draft because:

- I have not run cranky since I don't have it installed
- This could be done better, for example by only showing the window after the first frame has already been painted
- Ultimately I just wanted to submit this as a first step, even if it's not yet perfect.

### CODE REVIEW

**Comment by emilk** [comment]
> I have not run cranky since I don't have it installed

`cargo install cranky`

**Comment by emilk** [comment]
Thanks for working on this - this seems like a promising avenue to explore!

**Comment by LoganDark** [comment]
> > I have not run cranky since I don't have it installed
> 
> `cargo install cranky`

I think it's actually `cargo install cargo-cranky`.

Alright, it passes, and I also ran `cargo fmt` (which I previously forgot).

> Thanks for working on this - this seems like a promising avenue to explore!

I'm going to see if I can delay this until the first frame is already painted. That could eliminate the white flash altogether. A lot of the overhead seems to be glutin context creation which takes 500ms at best (iGPU) or 779ms at worst (dGPU).

**Comment by LoganDark** [comment]
The window being shown is now properly delayed until the first frame is *naturally* rendered. The rendering is not forced / done prematurely, either.

This also makes it possible to do something that you could not do before: keep the window completely hidden by calling `Frame.set_visible(false)` on the first update, in which case it will not be shown at any point.

There is still a very very brief (1 frame?) white flash. Maybe we could show the app on the second frame instead :^).

Marking as ready for review.

File: crates/eframe/src/native/epi_integration.rs [hunk 1]
```diff
@@ -341,6 +343,12 @@ impl EpiIntegration {
         app.post_rendering(window_size_px, &self.frame);
     }
 
+    pub fn post_present(&mut self, window: &winit::window::Window) {
+        if let Some(visible) = self.frame.output.visible.take() {
+           window.set_visible(visible);
```
Comment (emilk) [inline crates/eframe/src/native/epi_integration.rs:348]:
```suggestion
        if let Some(visible) = self.frame.output.visible.take() {
            window.set_visible(visible);
```

File: crates/eframe/src/native/run.rs [hunk 2]
```diff
@@ -330,8 +330,9 @@ mod glow_integration {
             };
             let window_settings = epi_integration::load_window_settings(storage);
 
-            let window_builder =
-                epi_integration::window_builder(native_options, &window_settings).with_title(title);
+            let window_builder = epi_integration::window_builder(native_options, &window_settings)
+                .with_title(title)
+                .with_visible(false);
```
Comment (emilk) [inline crates/eframe/src/native/run.rs:335]:
```suggestion
                .with_visible(false); // Keep hidden until we painted something. See https://github.com/emilk/egui/pull/2279
```

File: crates/eframe/src/native/run.rs [hunk 3]
```diff
@@ -692,6 +695,7 @@ mod wgpu_integration {
             let window_settings = epi_integration::load_window_settings(storage);
             epi_integration::window_builder(native_options, &window_settings)
                 .with_title(title)
+                .with_visible(false)
```
Comment (emilk) [inline crates/eframe/src/native/run.rs:698]:
```suggestion
                .with_visible(false) // Keep hidden until we painted something. See https://github.com/emilk/egui/pull/2279
```

File: crates/eframe/src/native/epi_integration.rs [hunk 4]
```diff
@@ -146,7 +146,7 @@ pub fn handle_app_output(
         fullscreen,
         drag_window,
         window_pos,
-        visible,
+        visible: _,
```
Comment (emilk) [inline crates/eframe/src/native/epi_integration.rs:149]:
```suggestion
        visible: _, // handled in post_present
```

File: crates/eframe/src/native/epi_integration.rs [hunk 5]
```diff
@@ -325,7 +324,10 @@ impl EpiIntegration {
             if app_output.close {
                 self.close = app.on_close_event();
             }
+            let visible = app_output.visible;
             handle_app_output(window, self.egui_ctx.pixels_per_point(), app_output);
+            // TODO(logandark): find a nicer way to do this
+            self.frame.output.visible = visible;
```
Comment (emilk) [inline crates/eframe/src/native/epi_integration.rs:330]:
```suggestion
            self.frame.output.visible = app_output.visible; // this is handled by post_present
            handle_app_output(window, self.egui_ctx.pixels_per_point(), app_output);
```

**Review by emilk — APPROVED** [review]

**Pull request merged**